### ISSUE TITLE
Basic implementation of spotlights systems

### ISSUE THREAD

**Issue by venix12** [ppy/osu-web#8736]
This issue is proposal of implementation of basic spotlights sytems into the osu-web, I'm willing to work on it, putting a draft in here, open for improvements.

### Basic infrastructure
- Add spotlights marker on spotlighted beatmapsets, and allow to filter by it (similar to how FA tags work). Those would be linked to URL from spotlights table mentioned below, or remain unclickable for legacy spotlights/charts implementations.
- Create ``spotlights`` table, which would contain fields such as title, gamemode, URL (such as newspost, wiki or something), and start/end dates beatmappack id (?) for modern (playlist-based) spotlights seasons along with ``spotlights_playlists`` pivot table where the lazer playlists would be attached to certain spotlights season.
- Create frontend for ``spotlights`` where the data would be displayed along with a list of atttached playlist lobbies.

This kind of implementation could make a base for creating a spotlights leaderboard in the future.

### Some ideas (would be nice to have but not really high priority)
- Some form of "subscribtion" - a person could follow spotlights and get a notification when new lobbies open (are attached to spotlights season).
- Admin panel of some sort to manage that all - probably can be just done via inserts/commands, would still be more handy to use, i guess.

**Comment by peppy** [comment]
I think a lot of what is asked for here should potentially be added to the playlist system itself, ie. allowing any playlist to have an attached description, URL, etc.

Maybe we could also implement a "seasons" system where multiple playlists could be grouped together (implement the backend for spotlights but eventually let others use this?)

**Comment by nanaya** [comment]
The title says "Basic implementation" so is it a different system from the lazer playlist's spotlight flag?

**Comment by peppy** [comment]
The goal here is to give spotlights more of an official presence as they expand in scope (ie. more users are interacting with the system). Making it visible on a beatmap's page (and potentially.. the beatmap listing as proposed), as well as allowing better linking of metadata so players can get more info on the spotlight/playlist they are playing on.

To get some context you can check the latest news post regarding the system usage (https://osu.ppy.sh/home/news/2022-02-15-beatmap-spotlights-season-5-winter-2022). At very least I'd want to see the ability for us to add descriptive text (maybe full markdown?) so spotlight ranking pages (and in-game display) could give users the same amount of detail as the news post. Or just link to the news post if that's simpler.

Basically it's very hard for a user to understand what spotlights are right now unless they follow news posts closely. Given that playlists/spotlights are an area I really hope to build on in the future expanding the general flexibility of the system seems like a good next step.

**Comment by nanaya** [comment]
So... extending current system and not a basic implementation? Can use a better issue title at least I think.

**Comment by venix12** [comment]
@peppy  since I'm gonna work on the seasons thing now, i wanted to ask, should there be an interOp API endpoint or something to make assigning certain rooms to certain seasons (assuming the ``Season`` creation can go straight into database) a thing in the script you're working on? https://github.com/ppy/osu-web/issues/8927#issuecomment-1131277619

The workflow could be pretty much:
1. a ``Season`` is created
2. script is hitting interOp endpoint along with ``season_id`` and ``room_id``, which then proceeds:
   - ``Room`` is marked as spotlight
   - ``Room`` items (beatmaps) are marked with spotlights tag and reindexed
   
That kind of solution seems pretty handy to manage all this stuff.

**Comment by venix12** [comment]
Thought you wanted to use it in an external script, that's why i suggested the interOp, did the stance change since and therefore i should make it a website UI thing instead?🤔

https://github.com/ppy/osu-web/issues/8927#issuecomment-1131277619
> If it makes more sense, we could implement this as an api endpoint and skip requiring the UI overhead. Or if it's something which is going to involve a manual switch from our end (ie. also flagging the playlists in lazer as spotlight), it can potentially be automated as part of that process.

**Comment by peppy** [comment]
The implementation with a `Season` table sounds best from a flexibility perspective.

I'll get to adding a command to do the management as soon as I find the bandwidth (right now making the changes manually is lower effort due to how rare the requests come in). But feel free to continue pushing forward with the surrounding implementation and I'll adapt as required.

**Comment by venix12** [comment]
ok, so just to clarify, should i go with ``season_id`` on ``Room`` or with separate ``SeasonRooms`` table to allow one-to-many relationship between rooms and seasons?

**Comment by peppy** [comment]
The second sounds better. Or if the list of rooms is decided at the start of the season and never changed, a CSV column could work fine too. Whatever is simpler.

**Comment by nanaya** [comment]
this one is all done, right?
